feat(core): add ng generate schematic to convert to inject

crisbeto opened the pull request:
Adds the new `ng generate @angular/core:inject-migration` schematic that will convert existing code from constructor-based injection to injection using the `inject` function. The migration also has a few options that should help reduce compilation errors.

This migration is slightly different than our usual ones in that it may have to update entire class or constructor declarations. We don't go through the `ts.factory.update*` APIs for this, because it can cause the entire declaration to be re-formatted. Instead, this migration tries to insert strings in a way that won't affect the user's formatting.

devversion commented on packages/core/schematics/ng-generate/inject-migration/README.md:
```diff
@@ -0,0 +1,118 @@
+export class MyComp {
+  private service = inject(MyService);
+
+  /** Inserted by Angular inject() migration for backawards compatibility */
```
```suggestion
  /** Inserted by Angular inject() migration for backwards compatibility */
```

devversion commented on packages/core/schematics/ng-generate/inject-migration/README.md:
```diff
@@ -0,0 +1,118 @@
+  private service = inject(MyService);
+
+  /** Inserted by Angular inject() migration for backawards compatibility */
+  constructor(...args: unknown[]);
```
Is it worth adding a TODO here for people, or a `UnknownForMigrationTODO` type

crisbeto replied:
Hmm, I'm on the fence since this is opt in.

devversion replied:
not feeling strongly, so leaving that to you

devversion commented on packages/core/schematics/ng-generate/inject-migration/README.md:
```diff
@@ -0,0 +1,118 @@
+
+@Component({})
+export class MyComp {
+  // Note the `!` at the end.
```
Do we insert a TODO here?

crisbeto replied:
I'm not sure about it. I wanted to keep it minimal given that it's an opt-in configuration.

devversion replied:
Fair enough. I think a TODO would be useful if it's easy to add

devversion commented on packages/core/schematics/ng-generate/inject-migration/analysis.ts:
```diff
@@ -0,0 +1,223 @@
+    if (ts.isClassDeclaration(node)) {
+      const decorators = getAngularDecorators(localTypeChecker, ts.getDecorators(node) || []);
+      const supportsDI = decorators.some(
+        (dec) => DECORATORS_SUPPORTING_DI.has(dec.name) && dec.moduleName === '@angular/core',
```
nit: isn't it already checked to be `@angular/core`?

crisbeto replied:
Ah yeah, initially I wanted to avoid the type checker completely and this is a leftover.

eneajaho commented on packages/core/schematics/ng-generate/inject-migration/README.md:
````diff
@@ -0,0 +1,118 @@
+
+**After:**
+```typescript
+import { Component, Inject, Optional, inject } from '@angular/core';
````
```suggestion
import { Component, inject } from '@angular/core';
```

eneajaho commented on packages/core/schematics/ng-generate/inject-migration/README.md:
````diff
@@ -0,0 +1,118 @@
+
+**After:**
+```typescript
+import { Component } from '@angular/core';
````
```suggestion
import { Component, inject } from '@angular/core';
```

eneajaho commented on packages/core/schematics/ng-generate/inject-migration/README.md:
````diff
@@ -0,0 +1,118 @@
+
+**After:**
+```typescript
+import { Component, Inject, Optional, inject } from '@angular/core';
````
```suggestion
import { Component, inject } from '@angular/core';
```

devversion commented on packages/core/schematics/ng-generate/inject-migration/analysis.ts:
```diff
@@ -0,0 +1,223 @@
+ * @param declaration Function in which to search for unused parameters.
+ * @param localTypeChecker Type checker scoped to the file in which the function was declared.
+ */
+export function getUnusedParameters(
```
What about parameters used inside the class, by e.g. being marked as `public`?

crisbeto replied:
If the parameter has a modifier, we always emit it. We only use this to determine if we need to add variables to the constructor.

eneajaho reviewed (commented):
Hi @crisbeto 
I saw that generics case is not supported, for example: `ElementRef<HTMLDivElement>`

Because `inject()` doesn't infer the type correctly (check this issue: https://github.com/angular/angular/issues/53894) I guess, we also need to pass the whole type as a generic to inject. 

```ts
el = inject<ElementRef<HTMLDivElement>>(ElementRef<HTMLDivElement>)
```

devversion commented on packages/core/schematics/ng-generate/inject-migration/analysis.ts:
```diff
@@ -0,0 +1,223 @@
+ * @param node Node for which to get the indentation level.
+ */
+export function getNodeIndentation(node: ts.Node): string {
+  const fullText = node.getFullText();
```
Could potentially be simplified by using `getCharacterAndLineOfNode(..).col` or something (not sure the exact name of the helper from TypeScript)

crisbeto replied:
Hmm, I think that using `getLineAndCharacterOfPosition` might not help much here. This function isn't getting the number of indents, but it gets the actual indent string. This makes it easier to deal with different line endings. Initially I was trying to do it with the column number as well, but more some reason doing `sourceFile.getFullText` wasn't super reliable. I'll keep it as is for now.

devversion replied:
I'm not sure I follow what this function gets you. I thought it gets the indentation of a node. e.g.

```
constructor(
 ____<my-node>
)
```

where the function would then return a string representing the space before `<my-node`. `getLineAndCharacterOfPosition` can get you that if you use `' '.repeat(position.col)`

crisbeto replied:
Yep, exactly. For a node like `    private foo: number`, it would get `    `.  We can't do `' '.repeat(position.col)`, because the file could be using tabs for indentation.

devversion replied:
Ahh, that's a good point.

devversion commented on packages/core/schematics/ng-generate/inject-migration/analysis.ts:
```diff
@@ -0,0 +1,223 @@
+    return node.types.some(isNullableType);
+  }
+
+  return false;
```
You can also use the type checker to check if it's nullable— but maybe this is intended to be just AST based

crisbeto replied:
I'll double check it, but my assumption was that it might not work if it can't resolve types coming from other files.

devversion replied:
Hmm. It might end up being the same, if the type is defined outside of the file. Although inside files, it would be more correct. Seems minor to not bother if that's the case

devversion commented on packages/core/schematics/ng-generate/inject-migration/migration.ts:
```diff
@@ -0,0 +1,473 @@
+  }
+
+  const sourceFile = node.getSourceFile();
+  const unusedParameters = getUnusedParameters(constructor, localTypeChecker);
```
Maybe this helper should be named `getUnusedInConstructorParameters`?

devversion commented on packages/core/schematics/ng-generate/inject-migration/migration.ts:
```diff
@@ -0,0 +1,473 @@
+  // Since the value gets passed through directly anyway, we generate the call using a placeholder
+  // which we then replace with the raw text of the `TypeNode`.
+  const injectRef = tracker.addImport(param.getSourceFile(), 'inject', moduleName);
+  const args: ts.Expression[] = [ts.factory.createIdentifier(PLACEHOLDER)];
```
why can't we use the raw text of `TypeNode` here already?

crisbeto replied:
Because it's a bit of a mix of string manipulation and creating an AST.

devversion replied:
I'm not sure I 100% follow, but it doesn't matter much, so LGTM

crisbeto replied:
Re-reading your comment now and I think I understand the question. If I remember correctly, TS does some sanitization of identifiers. I was a bit afraid that it might mangle a `TypeNode` that looks something like `Foo<Bar<Baz>>`.

devversion replied:
IIRC it only does this when using a "generated identifier" via e.g. `ts.createUniqueIdentifier`

devversion commented on packages/core/schematics/ng-generate/inject-migration/migration.ts:
```diff
@@ -0,0 +1,473 @@
+}
+
+/**
+ * Removes the parameters from a constructor.
```
Can you expand the comment here why this is such a complex logic?

crisbeto replied:
It was in a comment a bit further down, but I've moved it up.

devversion commented on packages/core/schematics/ng-generate/inject-migration/migration.ts:
```diff
@@ -0,0 +1,473 @@
+ * @param sourceFile File for which to create the type checker.
+ */
+function getLocalTypeChecker(sourceFile: ts.SourceFile) {
+  const options: ts.CompilerOptions = {noEmit: true};
```
Might even speed up things to add `skipLibCheck` etc.

devversion reviewed (approved):
LGTM overall. Just some minor comments

crisbeto committed:
feat(core): add ng generate schematic to convert to inject

Adds the new `ng generate @angular/core:inject-migration` schematic that will convert existing code from constructor-based injection to injection using the `inject` function. The migration also has a few options that should help reduce compilation errors.

This migration is slightly different than our usual ones in that it may have to update entire class or constructor declarations. We don't go through the `ts.factory.update*` APIs for this, because it can cause the entire declaration to be re-formatted. Instead, this migration tries to insert strings in a way that won't affect the user's formatting.

crisbeto commented:
Good point @eneajaho, I've added it.

Thank you for the quick review @devversion 🎉

eneajaho reviewed (commented):
All license links in the top comment point to: https://angular.io/license
Should we move them to: https://angular.dev/license

crisbeto commented:
I looked into changing the licenses a few weeks ago, but it turned out to be kinda huge so I decided to leave it for later: https://github.com/angular/angular/pull/56379

rainerhahnekamp commented:
Apologies for the question, but could you clarify why we are getting a migrator?

Did I miss a discussion or statement indicating that inject is now the preferred approach?

The last information I saw suggested that we could use any method we preferred, and that there would be a wait for the outcome of the standardization process of the parameter decorators.

Since the inject vs. constructor topic is quite polarizing, it might be helpful to release a statement before this change is merged.

crisbeto commented:
@rainerhahnekamp both constructor-based injection and `inject` are supported right now and will continue being supported in the foreseeable future. We're getting this migrator, because:
1. We've seen that there's similar things popping up both internally and externally so there's a clear need for it.
2. We had some requests for it internally so it made sense to just open-source this so that all Angular users can use it.
3. Given that standard decorators don't support function parameters, there may be projects out there that want to switch to `inject` so that they can turn off experimental decorators.

rainerhahnekamp commented:
Thanks @crisbeto, that was the answer I was looking for!

atscott commented:
This PR was merged into the repository by commit fab673a1ddbca19ff9734f92a5ef0cc16be5708c.

The changes were merged into the following branches: main
